### PR TITLE
[SPARK-5213] [SQL] Pluggable SQL Parser Support

### PR DESCRIPTION
**Pull request by scwf**
based on #4015, we should not delete `sqlParser` from sqlcontext, that leads to mima failed. Users implement dialect to give a fallback for `sqlParser`  and we should construct `sqlParser` in sqlcontext according to the dialect
`protected[sql] val sqlParser = new SparkSQLParser(getSQLDialect().parse(_))`

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 1]
```diff
@@ -132,17 +174,31 @@ class SQLContext(@transient val sparkContext: SparkContext)
   protected[sql] lazy val optimizer: Optimizer = DefaultOptimizer
 
   @transient
-  protected[sql] val ddlParser = new DDLParser(sqlParser.parse(_))
+  protected[sql] val ddlParser = new DDLParser((sql: String) => { getSQLDialect().parse(sql) })
 
+  // kept sqlParser to make sure mima tests pass
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: Remove this comment? It's probably confused once this PR merged.

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 2]
```diff
@@ -45,6 +48,45 @@ import org.apache.spark.util.Utils
...
+ *-- Exception will be thrown and switch to dialect
+ *-- "sql" (for SQLContext) or 
+ *-- "hiveql" (for HiveContext)
+ * }}}
+ */
+private[spark] class DefaultDialect extends Dialect {
+  @transient
+  protected val sqlParser = {
+    val catalystSqlParser = new catalyst.SqlParser
+    new SparkSQLParser(catalystSqlParser.parse)
```
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: My bad, `sqlParser` should just be `SqlParser`, not the `SparkSQLParser`

**Comment by chenghao-intel** [comment]
Thank you @scwf for the fixing. :)

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 1 above)
Comment (scwf) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: ok

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 2 above)
Comment (scwf) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: i think it should be SparkSQLParser here, otherwise we can not parse cache command

Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: https://github.com/apache/spark/pull/5827/files#diff-131c27c6a1f59770d738b11f2a4755ecR180

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala [hunk 3]
```diff
@@ -132,17 +174,30 @@ class SQLContext(@transient val sparkContext: SparkContext)
...
 
   @transient
-  protected[sql] val ddlParser = new DDLParser(sqlParser.parse(_))
+  protected[sql] val ddlParser = new DDLParser((sql: String) => { getSQLDialect().parse(sql) })
 
   @transient
-  protected[sql] val sqlParser = {
-    val fallback = new catalyst.SqlParser
-    new SparkSQLParser(fallback.parse(_))
+  protected[sql] val sqlParser = new SparkSQLParser(getSQLDialect().parse(_))
```
Comment (scwf) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: @chenghao-intel this sqlparser actually will not be used for now, place here just to fix mima test

**Comment by scwf** [comment]
Retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala (hunk 3 above)
Comment (chenghao-intel) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: I think we'd better keep it, not just for the mima test, but also for the sub class of `Dialect`. e.g. we have to specify the `SparkSQLParser` for `HiveQLDialect`.

Comment (scwf) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala]: agree to keep it, and in dialect parser we should not use SparkSQLParser. Dialect give a fallback(string -> logicalplan) and we call it in sqlParser

**Comment by scwf** [comment]
retest this please

**Comment by scwf** [comment]
@marmbrus any comment here?

**Comment by rxin** [comment]
@scwf are you guys using this feature? I'm thinking about just removing it in Spark 2.0. 

@chenghao-intel  who wanted it in the first place no longer needs it.

**Comment by scwf** [comment]
@rxin, yes we used this and we implements a new sqlparser based on this interface to support ANSI tpcds sql.

**Comment by rxin** [comment]
What's different from the one in Spark master now? It would be great to contribute the parser changes back now we have a full fledged parser in Spark, and going towards more ANSI compatibility is definitely on the roadmap.

**Comment by scwf** [comment]
@rxin Our parser is a extended version of the `SqlParser`,  the main difference is that we add the support for subquery(both correlated and uncorrelated ),exists, in and some minor improvement such as grouping, top, cube/rollup. It support the tpcds generated ANSI sql syntax without any change.

I noticed that there are some PRs for these features, i will take a look at that PRs when i have time and see what i can do.

**Comment by rxin** [comment]
FYI we are going to remove this pluggability. It is extra overhead to maintain, and actually encourages projects to not contribute their improvements upstream, which is bad.

**Comment by scwf** [comment]
Actually we were trying to contribute this improvements, unfortunately the community do not want them for maintain(or compatibility with hive ql) reason in the past:).

I am glad that spark sql use a single parser such that people can make contributions and make it more and more powerful.

**Comment by rxin** [comment]
Yup thanks. That's why we are only removing it now :)

**Pull request closed without merging**